src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -255,8 +254,7 @@ public static function get_theme_data( $deprecated = array(), $options = array()
 			 * @param WP_Theme_JSON_Data $theme_json Class to access and update the underlying data.
 			 */
 			$theme_json      = apply_filters( 'wp_theme_json_data_theme', new WP_Theme_JSON_Data( $theme_json_data, 'theme' ) );
-			$theme_json_data = $theme_json->get_data();
-			static::$theme   = new WP_Theme_JSON( $theme_json_data );
+			static::$theme   = $theme_json->get_theme_json();
```

Comment:
There are some linting issues in the committed code, incorrect spacing before the equals signs here.

The main reason I'm looking at this though is that a new feature was merged in Gutenberg - https://github.com/WordPress/gutenberg/pull/57908. It was testing fine prior to the changes on these lines, but after the change there's an exception being thrown by this change (I bisected to find it).

To repro:
1. Make sure you have the latest core trunk and gutenberg trunk code checked out and built, you have the plugin active and are using twentytwentyfour as your theme
2. Add a file with the following content into `wp-content/themes/twentytwentyfour/styles/block-style-contrast.json`:
```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"title": "Variation A",
	"blockTypes": [ "core/group", "core/columns", "core/media-text" ],
	"styles": {
		"color": {
			"background": "var(--wp--preset--color--contrast)",
			"text": "var(--wp--preset--color--base)"
		}
	}
}
```
3. Load the site editor, and you should see an error:
```
Warning: Undefined array key "styleVariations" in /var/www/html/wp-includes/class-wp-theme-json.php on line 2465
Warning: Trying to access array offset on value of type null in /var/www/html/wp-includes/class-wp-theme-json.php on line 2465
```

If you then revert the change on this line(s), the error goes away.

I don't have much of an understanding why this causes an issue. It'd be great to understand why it's happening before taking any action. I'll continue looking into it, but any help appreciated, as this'll need to be fixed prior to the 6.6 beta.

Reply:
@kt-12 @joemcgill @oandregal I think I have an understanding of what's happening, but worth noting I'm not an expert with these theme json classes.

The gutenberg codebase hooks into the `wp_theme_json_data_theme` filter, but can return a `WP_Theme_JSON_Gutenberg` from the filter. Before your change, it didn't really matter as core would construct a `WP_Theme_JSON` from that, but now I think the change can mean core holding on to a `WP_Theme_JSON_Gutenberg` reference. This might be leading to some incompatibilities in the data structures. At some point the newer `WP_Theme_JSON_Gutenberg` data is being processed by an older `WP_Theme_JSON` class. I'm not sure exactly where that happens, still investigating.

The core backport for the feature (https://github.com/WordPress/wordpress-develop/pull/6662) does resolve the issue, as it updates `WP_Theme_JSON` to be compatible.

But I think there are question marks over what happens when the gutenberg plugin tries to implement new features in the future that extend the theme json.

Reply:
Thanks @talldan. We had this reported to the related Trac ticket and have created [this follow-up PR](https://github.com/WordPress/wordpress-develop/pull/6626) to address the issue. If you could test it and confirm that it fixes the issue, I can get it committed.

> But I think there are question marks over what happens when the gutenberg plugin tries to implement new features in the future that extend the theme json.

I think this is a very good question. I've had some discussion with @tellthemachines and @oandregal about making the `WP_Theme_JSON_` related classes read only in the WP-dev repo and sync them from the GB repo the way we sync other packages, which is one way to solve these incompatibilities, but we've not confirmed that approach. I need to open a new issue for us to consider those changes.

Reply:
I finally got some time to look at this today. Sharing what I know.

The flow:

| Steps | Core | Gutenberg plugin |
| --- |  --- | --- |
| 1.    | WP_Theme_JSON triggers the `wp_theme_json_data_theme` filter. | |
| 2.    |                                                               | `gutenberg_resolve_style_variations_*` runs as a filter callback |
| 3.    |                                                               | Gutenberg builds [its own](https://github.com/WordPress/gutenberg/blob/044a05f1486f2631675ac3a98ddd64018c63e574/lib/block-supports/block-style-variations.php#L305) `WP_Theme_JSON_Data_Gutenberg` which uses to validate the data (`update_with`) and then returns. |
| 4.    | WP_Theme_JSON receives the validated data.                    | |

At 4, core receives some data and it expects it to be validated by the consumer of the filter (this is, the consumer uses `update_with`). While Gutenberg [validates the data](https://github.com/WordPress/gutenberg/blob/f7a3c656511bf93c0e28a388ab461a1f9abbf787/lib/block-supports/block-style-variations.php#L299) properly, it uses different metadata/schema than core does.

In the particular case that Dan raised, Gutenberg was able to get more style variations than core had:

| Variations detected by Gutenberg       | Variations detected by core    |
| -------------------------------------- | ------------------------------ |
| For block core/columns: variation-a    |                                |
| For block core/group: variation-a      |                                |
| For block core/media-text: variation-a |                                |
| For block core/button: outline         | For block core/button: outline |
| For block core/image: rounded          | For block core/image: rounded  |
| For block core/quote: plain            | For block core/quote: plain    |

At the point of processing the variations, core is unable to find the first three variations, hence the `PHP Warning:  Undefined array key "styleVariations" in /var/www/src/wp-includes/class-wp-theme-json.php on line 2445` raised [here](https://github.com/WordPress/wordpress-develop/blob/5115c7ac80d7260b73d467f784b790587d61210b/src/wp-includes/class-wp-theme-json.php#L2603). Before this PR, core did an extra validation step, getting rid of the _extraneous_ data before processing it, so that never happened.

There's a few layers to this issue, and it only happens because Gutenberg is re-triggering [the same filters](https://github.com/WordPress/gutenberg/blob/f7a3c656511bf93c0e28a388ab461a1f9abbf787/lib/class-wp-theme-json-resolver-gutenberg.php#L257-L258) than core (`wp_theme_json_data_theme`). The consumers of the filters run twice (one for core and the second time for the plugin). It's the first time that it fails because the metadata/schemas are different.

Unlike this [other issue](https://github.com/WordPress/wordpress-develop/pull/6626#issuecomment-2146277239), this is very Gutenberg-specific because it is the result of modifying the metadata/schema of the theme.json. I don't have any more time today to look into solutions, but wanted to share what I know about as soon as possible. I'll do some more testing in the next days to assess the severity and what can we do.

Reply:
For a different issue, I ended up proposing a change to how "block style variations" defined by the theme are registered at https://github.com/WordPress/wordpress-develop/pull/6756 By doing that, the error reported by Dan no longer happens.

Reply:
Thanks for looking into this @oandregal. It still seems like this is very similar to https://core.trac.wordpress.org/ticket/61112, but seems like the problem is that because core is not converting the `WP_Theme_JSON_Data_Gutenberg` object back into a `WP_Theme_JSON_Data` object, the incompatibilities are possible.

It seems wrong that Gutenberg would return an object from these filters that are not compatible with `WP_Theme_JSON_Data` objects, but we may be able to simply update the fix proposed in #6626 to check for whether the returned object is a `WP_Theme_JSON_Data` object, and reprocess if not, rather than just if it a `get_theme_json` method exists on the class.

As a matter of process, I think following up on this whole issue should be done in as a part of https://core.trac.wordpress.org/ticket/61112 so this doesn't get lost, so I'm going to summarize this issue on that ticket and update the PR with the new approach I just proposed.

Reply:
Updated the approach in https://github.com/WordPress/wordpress-develop/pull/6626 that should address the exception @talldan describes, though better ways of extending these objects in the Gutenberg repo are still worth investigating further.